From 994165f87de0fde103de3cbafe9041f1f2addab5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Markku=20Ahvenj=C3=A4rvi?= <mankku@gmail.com>
Date: Thu, 11 Jun 2026 09:34:45 +0300
Subject: [PATCH 2/2] pkvm: x86: Fix unhandled #VE exceptions
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Certain VMMs, namely crosvm, trap the OPRegion instead of having plain RAM.
This causes #VE for accesses. Unfortunately #VE exceptions handler only
supports a handful of instructions. Add quirks for operations that
exception handler is unable to handle.

Signed-off-by: Markku Ahvenjärvi <mankku@gmail.com>
---
 drivers/gpu/drm/i915/display/intel_bios.c     | 15 +++++++++++-
 drivers/gpu/drm/i915/display/intel_opregion.c | 23 ++++++++++++++++++-
 2 files changed, 36 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/display/intel_bios.c b/drivers/gpu/drm/i915/display/intel_bios.c
index bed485374ab0..d4d58c8237cd 100644
--- a/drivers/gpu/drm/i915/display/intel_bios.c
+++ b/drivers/gpu/drm/i915/display/intel_bios.c
@@ -2963,6 +2963,19 @@ static const struct bdb_header *get_bdb_header(const struct vbt_header *vbt)
 	return _vbt + vbt->bdb_offset;
 }
 
+static int check_vbt_signature(const struct vbt_header *vbt)
+{
+#if IS_ENABLED(CONFIG_PKVM_GUEST)
+	char signature[4] = { 0 };
+
+	memcpy_fromio(signature, vbt->signature, 4);
+
+	return memcmp(signature, "$VBT", 4);
+#else
+	return memcmp(vbt->signature, "$VBT", 4);
+#endif
+}
+
 /**
  * intel_bios_is_valid_vbt - does the given buffer contain a valid VBT
  * @display:	display device
@@ -2985,7 +2998,7 @@ bool intel_bios_is_valid_vbt(struct intel_display *display,
 		return false;
 	}
 
-	if (memcmp(vbt->signature, "$VBT", 4)) {
+	if (check_vbt_signature(vbt)) {
 		drm_dbg_kms(display->drm, "VBT invalid signature\n");
 		return false;
 	}
diff --git a/drivers/gpu/drm/i915/display/intel_opregion.c b/drivers/gpu/drm/i915/display/intel_opregion.c
index ff11836459de..65d7016bca13 100644
--- a/drivers/gpu/drm/i915/display/intel_opregion.c
+++ b/drivers/gpu/drm/i915/display/intel_opregion.c
@@ -1141,6 +1141,27 @@ bool intel_opregion_vbt_present(struct intel_display *display)
 	return true;
 }
 
+static const void *intel_opregion_vbt_dup(const void *src, size_t len, gfp_t gfp)
+{
+#if IS_ENABLED(CONFIG_PKVM_GUEST)
+	/* A quirk to handle situations where VMM models the region as MMIO.
+	 * kmemdup generates instructions that are not supported by the #VE
+	 * exception handler.
+	 */
+	void *dst;
+
+	dst = kmalloc(len, gfp);
+	if (!dst)
+		return NULL;
+
+	memcpy_fromio(dst, src, len);
+
+	return dst;
+#else
+	return kmemdup(src, len, GFP_KERNEL);
+#endif
+}
+
 const void *intel_opregion_get_vbt(struct intel_display *display, size_t *size)
 {
 	struct intel_opregion *opregion = display->opregion;
@@ -1151,7 +1172,7 @@ const void *intel_opregion_get_vbt(struct intel_display *display, size_t *size)
 	if (size)
 		*size = opregion->vbt_size;
 
-	return kmemdup(opregion->vbt, opregion->vbt_size, GFP_KERNEL);
+	return intel_opregion_vbt_dup(opregion->vbt, opregion->vbt_size, GFP_KERNEL);
 }
 
 bool intel_opregion_headless_sku(struct intel_display *display)
-- 
2.49.0

